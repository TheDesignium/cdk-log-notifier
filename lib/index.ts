import * as cdk from 'aws-cdk-lib';
import * as lambda from 'aws-cdk-lib/aws-lambda';
import * as logs from 'aws-cdk-lib/aws-logs';
import * as logsDestinations from 'aws-cdk-lib/aws-logs-destinations';
import * as constructs from 'constructs';
import * as path from 'path';

export interface LogNotifierAttributes {
  destinationFunctionArn: string;
  filterPattern: logs.IFilterPattern;
}

export interface LogNotifier extends LogNotifierAttributes {
  watch(logGroup: logs.ILogGroup): void;
};

abstract class LogNotifierImpl extends cdk.Resource implements LogNotifier {
  abstract readonly filterPattern: logs.IFilterPattern;
  protected abstract readonly handleLogFunc: lambda.IFunction;
  get destinationFunctionArn(): string {
    return this.handleLogFunc.functionArn;
  }
  get attributes(): LogNotifierAttributes {
    return {
      destinationFunctionArn: this.destinationFunctionArn,
      filterPattern: this.filterPattern,
    };
  }
  watch(logGroup: logs.LogGroup) {
    new logs.SubscriptionFilter(this, `${logGroup.node.id}SubscriptionFilter`, {
      destination: new logsDestinations.LambdaDestination(this.handleLogFunc),
      filterPattern: this.filterPattern,
      logGroup,
    });
  }
}

export interface LogNotifierProps {
  dateTimeFormat?: Intl.DateTimeFormat;
  filterPattern: logs.IFilterPattern;
  slackIncomingWebhookUrl: string;
}

class LogNotifierFacade extends LogNotifierImpl {
  filterPattern: logs.IFilterPattern;
  protected handleLogFunc: lambda.IFunction;
  constructor(
    scope: constructs.Construct,
    id: string,
    props: LogNotifierProps,
  ) {
    super(scope, id, { physicalName: id });

    const dateTimeFormat = props.dateTimeFormat ?? new Intl.DateTimeFormat('en-US', {
      timeZone: 'UTC',
      timeZoneName: 'short',
      month: 'numeric',
      day: 'numeric',
      hour: 'numeric',
      minute: 'numeric',
      second: 'numeric',
    });

    this.filterPattern = props.filterPattern;
    this.handleLogFunc = new lambda.Function(this, 'LogHandler', {
      code: lambda.Code.fromAsset(path.join(__dirname, 'lambda-log-handler')),
      runtime: lambda.Runtime.NODEJS_12_X,
      handler: 'index.handler',
      environment: {
        'RESOLVED_DATETIME_FORMAT_OPTIONS': JSON.stringify(dateTimeFormat.resolvedOptions()),
        'SLACK_INCOMING_WEBHOOK_URL': props.slackIncomingWebhookUrl,
      },
    });
  }
  static fromAttributes(scope: constructs.Construct, id: string, attrs: LogNotifierAttributes): LogNotifier {
    class LogNotifierFromAttributes extends LogNotifierImpl {
      filterPattern = attrs.filterPattern;
      protected handleLogFunc = lambda.Function.fromFunctionArn(scope, `${id}DestinationFunc`, attrs.destinationFunctionArn);
    }
    return new LogNotifierFromAttributes(scope, id, { physicalName: id });
  }
}

export const LogNotifier = LogNotifierFacade;
