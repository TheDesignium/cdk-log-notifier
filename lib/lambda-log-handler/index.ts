import * as zlib from 'zlib';
import * as https from 'https';
import type * as http from 'http';

if (process.env.SLACK_INCOMING_WEBHOOK_URL === undefined) throw 'SLACK_INCOMING_WEBHOOK_URL is not set.';
const slackIncomingWebhookUrl = process.env.SLACK_INCOMING_WEBHOOK_URL;

interface CloudWatchLogsEvent {
  awslogs: {
    data: string;
  };
}

interface LogData {
  logEvents: {
    id: string;
    timestamp: number;
    message: string;
  }[];
  logGroup: string;
  logStream: string;
  messageType: string;
}

function isLogData(v: any): v is LogData {
  if (!Array.isArray(v.logEvents)) return false;
  for (const logEvent of v.logEvents) {
    if (typeof logEvent.id !== 'string') return false;
    if (typeof logEvent.timestamp !== 'number') return false;
    if (typeof logEvent.message !== 'string') return false;
  }
  if (typeof v.logGroup !== 'string') return false;
  if (typeof v.logStream !== 'string') return false;
  if (typeof v.messageType !== 'string') return false;
  return true;
}

export const handler = async (event: CloudWatchLogsEvent) => {
  console.log('Received Event:', JSON.stringify(event));

  const logData = JSON.parse(zlib.gunzipSync(Buffer.from(event.awslogs.data, 'base64')).toString());

  if (!isLogData(logData)) {
    console.warn('Unregocgnizable message is received. Ignoreing.');
    return;
  }

  for (const logEvent of logData.logEvents) {
    let msg = logEvent.message;
    try {
      const jsonRegex = /{.*}/;
      const mayJson = msg.match(jsonRegex)?.[0];
      if (!mayJson) throw '';
      const json = JSON.parse(mayJson);
      msg = msg.replace(jsonRegex, JSON.stringify(json, null, 2));
    } catch (e) { }

    const resp = await fetch(slackIncomingWebhookUrl, {
      method: 'POST',
      headers: {
        'content-type': 'application/json',
      },
      body: JSON.stringify({
        "blocks": [
          {
            "type": "section",
            "text": {
              "type": "mrkdwn",
              "text": `${logData.logGroup}\n\`\`\`${msg}\n\`\`\``,
            },
          },
          {
            "type": "context",
            "elements": [
              {
                "type": "plain_text",
                "text": new Date(logEvent.timestamp).toLocaleString('ja-JP', {
                  timeZone: 'Asia/Tokyo',
                  month: 'numeric',
                  day: 'numeric',
                  hour: 'numeric',
                  minute: 'numeric',
                  second: 'numeric',
                }),
              },
              {
                "type": "mrkdwn",
                "text": `<https://ap-northeast-1.console.aws.amazon.com/cloudwatch/home?region=${process.env.AWS_REGION}#logsV2:log-groups/log-group/${encodeURIComponent(logData.logGroup)}/log-events/${encodeURIComponent(logData.logStream)}|See in CloudWatch>`,
              },
            ],
          },
        ],
      }),
    });

    if (!resp.ok) {
      console.error(`Posting message failed: status ${resp.status}; ${await resp.text()}`);
    }
  }
};

async function fetch(url: string, props: https.RequestOptions & { body: string }) {
  const res = await new Promise<http.IncomingMessage>((resolve, reject) => {
    const req = https.request(url, props, resolve);
    req.on('error', reject);
    req.end(props.body);
  });

  const bodyPromise = new Promise((resolve) => {
    let body = '';
    res.on('data', dat => body += dat);
    res.on('end', () => resolve(body));
  });

  return {
    ok: res.statusCode && res.statusCode >= 200 && res.statusCode < 300,
    status: res.statusCode,
    text: () => bodyPromise,
  };
}
