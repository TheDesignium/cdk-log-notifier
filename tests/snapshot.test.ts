import 'jest-cdk-snapshot';
import { expect, test } from 'vitest';
import { LogNotifier } from '../lib';
import * as logs from 'aws-cdk-lib/aws-logs';
import * as lambda from 'aws-cdk-lib/aws-lambda';
import { Stack } from 'aws-cdk-lib/core';

test('snapshot test', () => {
  const stack = new Stack();
  const lambda1 = new lambda.Function(stack, 'lambda', {
    runtime: lambda.Runtime.NODEJS_18_X,
    handler: 'index.handler',
    code: lambda.Code.fromInline('exports.handler = async function() { return "Hello, world!"; }'),
  });
  const logNotifier = new LogNotifier(stack, 'logNotifier', {
    filterPattern: logs.FilterPattern.allTerms('ERROR'),
    slackIncomingWebhookUrl: 'https://hooks.slack.com/test',
  });

  logNotifier.watch(lambda1.logGroup);

  expect(stack).toMatchCdkSnapshot({
    ignoreAssets: true,
  });
});
