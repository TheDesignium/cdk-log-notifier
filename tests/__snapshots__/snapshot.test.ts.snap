// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`snapshot test 1`] = `
{
  "Mappings": {
    "LatestNodeRuntimeMap": {
      "af-south-1": {
        "value": "nodejs20.x",
      },
      "ap-east-1": {
        "value": "nodejs20.x",
      },
      "ap-northeast-1": {
        "value": "nodejs20.x",
      },
      "ap-northeast-2": {
        "value": "nodejs20.x",
      },
      "ap-northeast-3": {
        "value": "nodejs20.x",
      },
      "ap-south-1": {
        "value": "nodejs20.x",
      },
      "ap-south-2": {
        "value": "nodejs20.x",
      },
      "ap-southeast-1": {
        "value": "nodejs20.x",
      },
      "ap-southeast-2": {
        "value": "nodejs20.x",
      },
      "ap-southeast-3": {
        "value": "nodejs20.x",
      },
      "ap-southeast-4": {
        "value": "nodejs20.x",
      },
      "ap-southeast-5": {
        "value": "nodejs20.x",
      },
      "ap-southeast-7": {
        "value": "nodejs20.x",
      },
      "ca-central-1": {
        "value": "nodejs20.x",
      },
      "ca-west-1": {
        "value": "nodejs20.x",
      },
      "cn-north-1": {
        "value": "nodejs18.x",
      },
      "cn-northwest-1": {
        "value": "nodejs18.x",
      },
      "eu-central-1": {
        "value": "nodejs20.x",
      },
      "eu-central-2": {
        "value": "nodejs20.x",
      },
      "eu-isoe-west-1": {
        "value": "nodejs18.x",
      },
      "eu-north-1": {
        "value": "nodejs20.x",
      },
      "eu-south-1": {
        "value": "nodejs20.x",
      },
      "eu-south-2": {
        "value": "nodejs20.x",
      },
      "eu-west-1": {
        "value": "nodejs20.x",
      },
      "eu-west-2": {
        "value": "nodejs20.x",
      },
      "eu-west-3": {
        "value": "nodejs20.x",
      },
      "il-central-1": {
        "value": "nodejs20.x",
      },
      "me-central-1": {
        "value": "nodejs20.x",
      },
      "me-south-1": {
        "value": "nodejs20.x",
      },
      "mx-central-1": {
        "value": "nodejs20.x",
      },
      "sa-east-1": {
        "value": "nodejs20.x",
      },
      "us-east-1": {
        "value": "nodejs20.x",
      },
      "us-east-2": {
        "value": "nodejs20.x",
      },
      "us-gov-east-1": {
        "value": "nodejs18.x",
      },
      "us-gov-west-1": {
        "value": "nodejs18.x",
      },
      "us-iso-east-1": {
        "value": "nodejs18.x",
      },
      "us-iso-west-1": {
        "value": "nodejs18.x",
      },
      "us-isob-east-1": {
        "value": "nodejs18.x",
      },
      "us-west-1": {
        "value": "nodejs20.x",
      },
      "us-west-2": {
        "value": "nodejs20.x",
      },
    },
  },
  "Resources": {
    "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aFD4BFC8A": {
      "DependsOn": [
        "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRoleDefaultPolicyADDA7DEB",
        "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRole9741ECFB",
      ],
      "Properties": {
        "Code": Any<Object>,
        "Handler": "index.handler",
        "Role": {
          "Fn::GetAtt": [
            "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRole9741ECFB",
            "Arn",
          ],
        },
        "Runtime": {
          "Fn::FindInMap": [
            "LatestNodeRuntimeMap",
            {
              "Ref": "AWS::Region",
            },
            "value",
          ],
        },
        "Timeout": 900,
      },
      "Type": "AWS::Lambda::Function",
    },
    "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRole9741ECFB": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRoleDefaultPolicyADDA7DEB": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "logs:PutRetentionPolicy",
                "logs:DeleteRetentionPolicy",
              ],
              "Effect": "Allow",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRoleDefaultPolicyADDA7DEB",
        "Roles": [
          {
            "Ref": "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRole9741ECFB",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "lambda8B5974B5": {
      "DependsOn": [
        "lambdaServiceRole494E4CA6",
      ],
      "Properties": {
        "Code": Any<Object>,
        "Handler": "index.handler",
        "Role": {
          "Fn::GetAtt": [
            "lambdaServiceRole494E4CA6",
            "Arn",
          ],
        },
        "Runtime": "nodejs20.x",
      },
      "Type": "AWS::Lambda::Function",
    },
    "lambdaLogRetention4CE10F74": {
      "Properties": {
        "LogGroupName": {
          "Fn::Join": [
            "",
            [
              "/aws/lambda/",
              {
                "Ref": "lambda8B5974B5",
              },
            ],
          ],
        },
        "ServiceToken": {
          "Fn::GetAtt": [
            "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aFD4BFC8A",
            "Arn",
          ],
        },
      },
      "Type": "Custom::LogRetention",
    },
    "lambdaServiceRole494E4CA6": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "logNotifierLogHandler893A9F2E": {
      "DependsOn": [
        "logNotifierLogHandlerServiceRole20A6D23C",
      ],
      "Properties": {
        "Code": Any<Object>,
        "Environment": {
          "Variables": {
            "RESOLVED_DATETIME_FORMAT_OPTIONS": "{"locale":"en-US","calendar":"gregory","numberingSystem":"latn","timeZone":"UTC","hourCycle":"h12","hour12":true,"month":"numeric","day":"numeric","hour":"numeric","minute":"2-digit","second":"2-digit","timeZoneName":"short"}",
            "SLACK_INCOMING_WEBHOOK_URL": "https://hooks.slack.com/test",
          },
        },
        "Handler": "index.handler",
        "Role": {
          "Fn::GetAtt": [
            "logNotifierLogHandlerServiceRole20A6D23C",
            "Arn",
          ],
        },
        "Runtime": {
          "Fn::FindInMap": [
            "LatestNodeRuntimeMap",
            {
              "Ref": "AWS::Region",
            },
            "value",
          ],
        },
      },
      "Type": "AWS::Lambda::Function",
    },
    "logNotifierLogHandlerServiceRole20A6D23C": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "logNotifierlambdaLogGroupSubscriptionFilterB8C78BD2": {
      "DependsOn": [
        "logNotifierlambdaLogGroupSubscriptionFilterCanInvokeLambda52A75F3B",
      ],
      "Properties": {
        "DestinationArn": {
          "Fn::GetAtt": [
            "logNotifierLogHandler893A9F2E",
            "Arn",
          ],
        },
        "FilterPattern": ""ERROR"",
        "LogGroupName": {
          "Fn::GetAtt": [
            "lambdaLogRetention4CE10F74",
            "LogGroupName",
          ],
        },
      },
      "Type": "AWS::Logs::SubscriptionFilter",
    },
    "logNotifierlambdaLogGroupSubscriptionFilterCanInvokeLambda52A75F3B": {
      "Properties": {
        "Action": "lambda:InvokeFunction",
        "FunctionName": {
          "Fn::GetAtt": [
            "logNotifierLogHandler893A9F2E",
            "Arn",
          ],
        },
        "Principal": "logs.amazonaws.com",
        "SourceArn": {
          "Fn::Join": [
            "",
            [
              "arn:",
              {
                "Ref": "AWS::Partition",
              },
              ":logs:",
              {
                "Ref": "AWS::Region",
              },
              ":",
              {
                "Ref": "AWS::AccountId",
              },
              ":log-group:",
              {
                "Fn::GetAtt": [
                  "lambdaLogRetention4CE10F74",
                  "LogGroupName",
                ],
              },
              ":*",
            ],
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
  },
}
`;
